KLR/NKI/Types.lean

```diff
@@ -0,0 +1,457 @@
+    match t.typ with
+    | .styp typ =>
+      let natParams : List String := List.map TypParam.name <| t.params.filter (Srt.isNat ∘ TypParam.sort)
+      let typParams : List String := List.map TypParam.name <| t.params.filter (Srt.isTyp ∘ TypParam.sort)
+      have hnat : natParams.length = TypParam.nnat t.params := List.length_map _
+      have htyp : typParams.length = TypParam.ntyp t.params := List.length_map _
+      hnat ▸ htyp ▸ typ.toDeBruijn natParams typParams
+    | .snat _ => none
+
+  def SNat.inferBinders (l : List TypParam) : SNat → List TypParam
+    | .param name =>
+      if name ∈ (l.filter (Srt.isNat ∘ TypParam.sort)).map TypParam.name then
+        l
+      else
+        l.concat ⟨name, .snat⟩
+    | .const _ => l
+
+  def STyp.inferBindersAux (l : List TypParam) : STyp → List TypParam
+    | .param name =>
+      if name ∈ (l.filter (Srt.isTyp ∘ TypParam.sort)).map TypParam.name then
+        l
+      else
+        l.concat ⟨name, .styp⟩
+    | .none | .bool | .int | .float | .string => l
+    | .tuple typs => l ++ (typs.map (STyp.inferBindersAux l)).flatten
+    | .tensor shapes _ => shapes.foldl SNat.inferBinders l
+    | .func dom ran => l ++ dom.inferBindersAux l ++ ran.inferBindersAux l
+
+  /-- Infer the list of type parameters -/
+  def STyp.inferBinders (t : STyp) : List TypParam :=
+    t.inferBindersAux []
+
+  /-- Infer the De Bruijn indexed version -/
+  def STyp.inferDeBruijn (t : STyp) : Option (NKI.STyp (TypParam.nnat t.inferBinders) (TypParam.ntyp t.inferBinders)) :=
+    Typ.toDeBruijn ⟨t.inferBinders, .styp t⟩
+
+  /- Parser for type annotations (`pt_` prefix for Py Types) -/
+
+  open Lean Elab Meta
+
+  /- SNat -/
+  declare_syntax_cat pt_nat
+  syntax ident : pt_nat
+  syntax num : pt_nat
+
+  /- DType -/
+  declare_syntax_cat pt_dtype
+  syntax "bfloat16" : pt_dtype
+  syntax "float8e3" : pt_dtype
+  syntax "float8e4" : pt_dtype
+  syntax "float8e5" : pt_dtype
+  syntax "float16" : pt_dtype
+  syntax "float32" : pt_dtype
+  syntax "float32r" : pt_dtype
+  syntax "int8" : pt_dtype
+  syntax "int16" : pt_dtype
+  syntax "int64" : pt_dtype
+  syntax "int32" : pt_dtype
+  syntax "uint8" : pt_dtype
+  syntax "uint16" : pt_dtype
+  syntax "uint32" : pt_dtype
+  syntax "uint64" : pt_dtype
+
+  /- STyp -/
+  declare_syntax_cat pt_typ
+
+  /- Primitive types -/
+  syntax "none" : pt_typ
+  syntax "bool" : pt_typ
+  syntax "int" : pt_typ
+  syntax "float" : pt_typ
+  syntax "string" : pt_typ
+
+  /- Type parameter -/
+  syntax ident : pt_typ
+
+  /- Tuple syntax -/
+  declare_syntax_cat pt_nat_tup
+  syntax pt_nat : pt_nat_tup
+  syntax pt_nat ", " pt_nat_tup : pt_nat_tup
+
+  declare_syntax_cat pt_typ_tup
+  syntax pt_typ : pt_typ_tup
+  syntax pt_typ ", " pt_typ_tup : pt_typ_tup
+
+  syntax "tuple[" pt_typ_tup "]" : pt_typ
+  syntax "tensor[" pt_dtype ", " "(" pt_nat_tup ")" "]" : pt_typ
+
+  syntax pt_typ " → " pt_typ : pt_typ
```

Comment:
nit: can we allow ascii as well, ->. I have an irrational allergy to unicode.

Reply:
Done!